pkg/cloud-init/cloud-init.go

```diff
@@ -120,6 +130,40 @@ func cloudInitUUIDFromVMI(vmi *v1.VirtualMachineInstance) string {
 	return string(vmi.Spec.Domain.Firmware.UUID)
 }
 
+func generateHostname(vmi *v1.VirtualMachineInstance, userData string) (string, error) {
+	const errPrefix = "cannot generate hostname:"
+
+	if vmi != nil && vmi.Spec.Hostname != "" {
+		return vmi.Spec.Hostname, nil
+	}
+
+	// This indicates that userdata is of YAML format.
+	// For more info: https://cloudinit.readthedocs.io/en/latest/topics/format.html
+	if !strings.HasPrefix(userData, "#cloud-config") && !strings.HasPrefix(userData, "Content-Type: text/cloud-config") {
+		if vmi == nil {
+			return "", fmt.Errorf("%s hostname was not defined in user data and vmi is nil", errPrefix)
+		}
+
+		return dns.SanitizeHostname(vmi), nil
+	}
+
+	// Remove comments
+	commentRegex := regexp.MustCompile("(?m)^#.*")
```

Comment:
Nit: adding a `\n` at the end would remove the entire line.
Also, is this really needed? I'd expect `yaml.Unmarshall()` to ignore comments, no?

Reply:
Nice idea! I've added `\n?` since comments don't have to end with a newline (e.g. if the comment is in the last line)

And yes, weirdly enough the unmarshalling fails on comment line

Reply:
It looks to me that [yaml v3](https://pkg.go.dev/gopkg.in/yaml.v3) should support comments: https://go.dev/play/p/IosnuGB_U-m

Here in the code v2 is used. Maybe give it a try?

Reply:
Good one!! Worked great!
Done.